### PR TITLE
Fix HTTP/3 ALPN

### LINKED ISSUE #55894: [HTTP/3] Update ALPN to h3
The latest draft has finalized "h3" as the ALPN identifier.

https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34#appendix-B.1

>    *  Removed draft version guidance; added final version string

https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34#section-3

>    A client MAY attempt access to a resource with an "https" URI by
>    resolving the host identifier to an IP address, establishing a QUIC
>    connection to that address on the indicated port (including
>    validation of the server certificate as described above), and sending
>    an HTTP/3 request message targeting the URI to the server over that
>    secured connection.  Unless some other mechanism is used to select
>    HTTP/3, **the token "h3" is used in the Application Layer Protocol
>    Negotiation (ALPN; see [RFC7301]) extension during the TLS handshake.**

https://datatracker.ietf.org/doc/html/draft-ietf-quic-http-34#section-11.1

>    This document creates a new registration for the identification of
>    HTTP/3 in the "Application Layer Protocol Negotiation (ALPN) Protocol
>    IDs" registry established in [RFC7301].
> 
>    The "h3" string identifies HTTP/3:
> 
>    Protocol:  HTTP/3
> 
>    Identification Sequence:  0x68 0x33 ("h3")
> 
>    Specification:  This document

### PR DESCRIPTION
**Pull request by CarnaViire**
Fixes #55894

----

Re #54726: I've taken a quick look on whether interop tests would start working after this. For `Public_Interop_Upgrade_Success`, it now works with LiteSpeed and Cloudflare servers and does not with Chromium one (error CONNECTION_IDLE). `Public_Interop_Upgrade_Success` still does not work at all, however the test itself seemed strange to me, but I didn't spend much time analyzing.

### CODE REVIEW

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs [hunk 1]
```diff
@@ -34,9 +34,7 @@ public Http3LoopbackServer(QuicImplementationProvider quicImplementationProvider
                     EnabledSslProtocols = options.SslProtocols,
                     ApplicationProtocols = new List<SslApplicationProtocol>
                     {
-                        new SslApplicationProtocol("h3-31"),
-                        new SslApplicationProtocol("h3-30"),
-                        new SslApplicationProtocol("h3-29")
+                        new SslApplicationProtocol("h3")
```
Comment (wfurt) [inline src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:37]: Should we add Http3 to

https://github.com/dotnet/runtime/blob/361184ff71b53ed8565bf8a7ecf3d4809c5ea689/src/libraries/System.Net.Security/ref/System.Net.Security.cs#L121-L122

(fine IMHO if we do it as follow-up)

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs [hunk 2]
```diff
@@ -21,9 +21,7 @@ namespace System.Net.Http
     internal sealed class Http3Connection : HttpConnectionBase
     {
         // TODO: once HTTP/3 is standardized, create APIs for this.
-        public static readonly SslApplicationProtocol Http3ApplicationProtocol29 = new SslApplicationProtocol("h3-29");
-        public static readonly SslApplicationProtocol Http3ApplicationProtocol30 = new SslApplicationProtocol("h3-30");
-        public static readonly SslApplicationProtocol Http3ApplicationProtocol31 = new SslApplicationProtocol("h3-31");
+        public static readonly SslApplicationProtocol Http3ApplicationProtocol = new SslApplicationProtocol("h3");
```
Comment (wfurt) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:24]: what do do with comment above?

**Review by wfurt — COMMENTED** [review]
Should we create test to verify that setting HTTP version to 3.0 will send correct/expected ALPN in handshake?

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs (hunk 1 above)
Comment (CarnaViire) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:37]: I was reluctant to bring public API change as we are now past Feature-Complete date.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs (hunk 2 above)
Comment (CarnaViire) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs:24]: Same as above - the comment is about adding public API to SslApplicationProtocol, so I left it there for now.

**Comment by CarnaViire** [comment]
> Should we create test to verify that setting HTTP version to 3.0 will send correct/expected ALPN in handshake?

I was observing transport (msquic) failures with ALPN_NEG_FAILURE when I've updated Http3Connection to "h3" but not yet updated Http3LoopbackServer, so existing tests kind-of check that implicitly. But I'll see how to create more explicit test.

**Comment by CarnaViire** [comment]
I am observing a lot of stress failures for HTTP/3 (but it's not like *all* of them fail). I don't know whether it was this change that regressed it or some other change. I would expect it to fail completely if it was this change, so I have doubts, but I will investigate.

**Comment by wfurt** [comment]
is there need to update asp.net as well?

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs (hunk 1 above)
Comment (geoffkizer) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:37]: We should just make sure we have a 7.0 issue filed on this. I don't think we care about it for 6.0.

**Review by geoffkizer — APPROVED** [review]

**Comment by geoffkizer** [comment]
Re this:

> does not with Chromium one (error CONNECTION_IDLE).

This seems super weird, any idea what's up here?

**Review by JamesNK — APPROVED** [review]
👍

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs (hunk 1 above)
Comment (CarnaViire) [reply src/libraries/Common/tests/System/Net/Http/Http3LoopbackServer.cs:37]: It's here https://github.com/dotnet/runtime/issues/1293

**Comment by CarnaViire** [comment]
Stress regression is unrelated, shows on main, created an issue for that https://github.com/dotnet/runtime/issues/56310

**Comment by CarnaViire** [comment]
> Should we create test to verify that setting HTTP version to 3.0 will send correct/expected ALPN in handshake?

@wfurt I wasn't able to come up with an idea on how to verify that explicitly. ALPN is hidden in QUIC handshake, isn't it? So I don't know how to intercept msquic to get that information... Do you have some ideas on how I could achieve that?

----

> > does not with Chromium one (error CONNECTION_IDLE).
>
> This seems super weird, any idea what's up here?

@geoffkizer It could be that the server just isn't working for some reason, but we need to investigate in #54726

----

> is there need to update asp.net as well?

@wfurt ASP.NET Core tests will need to be updated
https://github.com/dotnet/aspnetcore/blob/9da42b9fab4c61fe46627ac0c6877905ec845d5a/src/Servers/Kestrel/Transport.Quic/test/QuicTestHelpers.cs#L28  (cc @JamesNK)
But `QuicTransportOptions` does not seem to have any default value https://github.com/dotnet/aspnetcore/blob/8b30d862de6c9146f466061d51aa3f1414ee2337/src/Servers/Kestrel/Transport.Quic/src/QuicTransportOptions.cs  In stress tests I specified ALPN manually for Kestrel https://github.com/dotnet/runtime/pull/56215/files#diff-73d3348ab4b049ae3561a8faf2c2ea98d1a379e035ecb65ccf5f21801b41a07fR130 And as I don't see any ALPN_NEG_FAILURE there and success rate is same as in main, I guess that's ok

**Comment by wfurt** [comment]
I don't think we need to hold back for the test @CarnaViire. It would be _nice_. I see two possible strategies. QuicConnection has NegotiatedApplicationProtocol. And the handshake happens early on. So it seems like if you create QuicListener and do HttpRequest to it you should know - even if everything else fails. Alternatively we could update Loopback server and/or get access to it via some reflection.

**Comment by JamesNK** [comment]
> is there need to update asp.net as well?

We'll update our configurable ALPN to h3. The property will probably be removed for RC1.

**Review by wfurt — APPROVED** [review]
LGTM

**Pull request merged**